include/kvs_async_client.hpp

```diff
@@ -0,0 +1,548 @@
+#define SRC_INCLUDE_ASYNC_CLIENT_HPP_
+
+#include "common.hpp"
+#include "kvs.pb.h"
+#include "requests.hpp"
+#include "threads.hpp"
+#include "types.hpp"
+
+using TimePoint = std::chrono::time_point<std::chrono::system_clock>;
+
+struct PendingRequest {
+  TimePoint tp_;
+  Address worker_addr_;
+  pair<KeyRequest, Address> request_addr_pair_;
+};
+
+class KvsAsyncClient {
+ public:
+  /**
+   * @addrs A vector of routing addresses.
+   * @routing_thread_count The number of thread sone ach routing node
+   * @ip My node's IP address
+   * @tid My client's thread ID
+   * @timeout Length of request timeouts in ms
+   */
+  KvsAsyncClient(vector<UserRoutingThread> routing_threads, string ip,
+                 unsigned tid = 0, unsigned timeout = 10000) :
+      routing_threads_(routing_threads),
+      ut_(UserThread(ip, tid)),
+      context_(zmq::context_t(1)),
+      socket_cache_(SocketCache(&context_, ZMQ_PUSH)),
+      key_address_puller_(zmq::socket_t(context_, ZMQ_PULL)),
+      response_puller_(zmq::socket_t(context_, ZMQ_PULL)),
+      log_(spdlog::basic_logger_mt("client_log", "client_log.txt", true)),
+      timeout_(timeout) {
+    // initialize logger
+    log_->flush_on(spdlog::level::info);
+
+    // set class variables
+    bad_response_.set_response_id("NULL_ERROR");
+
+    std::hash<string> hasher;
+    seed_ = time(NULL);
+    seed_ += hasher(ip);
+    seed_ += tid;
+    log_->info("Random seed is {}.", seed_);
+
+    // bind the two sockets we listen on
+    key_address_puller_.bind(ut_.key_address_bind_address());
+    response_puller_.bind(ut_.response_bind_address());
+
+    pollitems_ = {
+        {static_cast<void*>(key_address_puller_), 0, ZMQ_POLLIN, 0},
+        {static_cast<void*>(response_puller_), 0, ZMQ_POLLIN, 0},
+    };
+
+    // set the request ID to 0
+    rid_ = 0;
+  }
+
+  ~KvsAsyncClient() {}
+
+ public:
+  /**
+   * Issue an async PUT request to the KVS for a certain lattice typed value.
+   */
+  void put_async(const Key& key, const string& payload,
+                 LatticeType lattice_type,
+                 const Address& response_addr = string()) {
+    KeyRequest request;
+    KeyTuple* tuple = prepare_data_request(request, key);
+    request.set_type(RequestType::PUT);
+    tuple->set_lattice_type(lattice_type);
+    tuple->set_payload(payload);
+
+    try_request(request, response_addr);
+  }
+
+  /**
+   * Issue an async GET request to the KVS for a cross causal lattice value.
+   */
+  void get_async(const Key& key, const Address& response_addr = string()) {
+    // we issue GET only when it is not in the pending map
+    if (pending_get_response_map_.find(key) ==
+        pending_get_response_map_.end()) {
+      KeyRequest request;
+      prepare_data_request(request, key);
+      request.set_type(RequestType::GET);
+
+      try_request(request, response_addr);
+    }
+  }
+
+  vector<pair<KeyResponse, Address>> receive_async(ZmqUtilInterface* kZmqUtil) {
```

Comment:
what's the alternative design in your mind?

Reply:
I guess I didn't understand why the pair has addresses in it? Is it because that's who's waiting for the response? But aren't we currently storing that in the `cover_map`?